client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/protocol/PutOutputStream.java

```diff
@@ -0,0 +1,113 @@
+package org.infinispan.client.hotrod.impl.protocol;
+
+import static org.infinispan.client.hotrod.impl.Util.await;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+
+public class PutOutputStream extends OutputStream {
+   private static final int BUFFER_SIZE = 8 * 1024;
+   private final BiFunction<ByteBuf, Boolean, CompletionStage<Void>> consumer;
+   private final ByteBufAllocator alloc;
+   private final AtomicInteger pendingWrites = new AtomicInteger();
+   private ByteBuf buf;
+   private volatile Throwable throwable;
+
+   public PutOutputStream(BiFunction<ByteBuf, Boolean, CompletionStage<Void>> consumer, ByteBufAllocator alloc) {
+      this.consumer = consumer;
+      this.alloc = alloc;
+   }
+
+   private void alloc() {
+      buf = alloc.buffer(BUFFER_SIZE, BUFFER_SIZE);
+   }
+
+   private void consume(ByteBuf buf, Boolean complete) {
+      var stage = consumer.apply(buf, complete)
+            .whenComplete((___, t) -> {
+               pendingWrites.decrementAndGet();
+               if (t != null) {
+                  throwable = t;
+               }
+            });
+      // Don't let more than 2 pending writes at any time
+      if (complete || pendingWrites.incrementAndGet() >= 2) {
```

Comment:
We could, but this would require additional threads on the server side, it might actually make it slower. The whole point of this pendingWrites is so we don't have too many requests sent to the server in a row (remember all of this is processed on the same event loop so they finish sequentially).

Reply:
Argh, yeah. I forgot this should come from the same loop.